src/TraceEvent/BPerf/BPerfEventSource.cs

```diff
@@ -0,0 +1,312 @@
+﻿//     Copyright (c) Microsoft Corporation.  All rights reserved.
+// This file is best viewed using outline mode (Ctrl-M Ctrl-O)
+//
+// This program uses code hyperlinks available as part of the HyperAddin Visual Studio plug-in.
+// It is available from http://www.codeplex.com/hyperAddin 
+// 
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.Diagnostics.Tracing.Parsers;
+using Microsoft.Diagnostics.Tracing.Parsers.Kernel;
+
+namespace Microsoft.Diagnostics.Tracing
+{
+    public sealed class BPerfEventSource : TraceEventDispatcher, IDisposable
+    {
+        private const int OffsetToExtendedData = 88; // offsetof(TraceEventNativeMethods.EVENT_RECORD*, ExtendedData);
+
+        private const ushort CompressionFormatXpressHuff = 4;
+
+        private const int BufferSize = 1024 * 1024 * 2;
+
+        private const int ReadAheadBufferSize = 1024 * 1024 * 1;
+
+        private static readonly Guid EventTraceGuid = new Guid("{68fdd900-4a3e-11d1-84f4-0000f80464e3}");
+
+        private readonly string btlFilePath;
+
+        private readonly long fileOffset;
+
+        private readonly byte[] workspace;
+
+        private readonly byte[] uncompressedBuffer;
+
+        private readonly byte[] compressedBuffer;
+
+        private readonly Dictionary<int, string> processNameForID = new Dictionary<int, string>();
+
+        private int eventsLost;
+
+        public BPerfEventSource(string btlFilePath)
+            : this(btlFilePath, 0)
+        {
+        }
+
+        public BPerfEventSource(string btlFilePath, long fileOffset)
+            : this(btlFilePath, fileOffset, new byte[BufferSize], new byte[BufferSize], new byte[BufferSize])
+        {
+        }
+
+        public BPerfEventSource(string btlFilePath, long fileOffset, byte[] workspace, byte[] uncompressedBuffer, byte[] compressedBuffer)
```

Comment:
Can you describe the uses of the different constructors here?  Specifically, why would you or would you not specify the buffers?

Reply:
Yup will do. The reason to specify buffers is to reduce allocations in a server context.

Reply:
Is there a reason to specify them to the constructor vs. allocating them as byte arrays that are private to the source?  In both cases, they are byte[] and so the same number of allocations will need to occur either way.  Do the sizes vary?

Reply:
Because one use case of this will be tools that already have these workspace buffers allocated associated with some context. And yes the sizes are fixed.